source/transactions/transactions.rst

```diff
@@ -744,24 +753,70 @@ format:
 
     { ok : 1, writeConcernError: {code: <Number>, errmsg : "..."} }
 
-Mongos Pinning for Sharded Transactions
----------------------------------------
+Sharded Transactions
+--------------------
+
+MongoDB 4.2 (maxWireVersion 8) introduces support for sharded transactions.
+Sharded transactions support all of the same features as single replica set
+transaction but introduce two new driver concepts: mongos pinning and the
+``recoveryToken`` field.
 
-The server supports sharded transactions starting in MongoDB 4.2 (
-maxWireVersion 8). The server requires that drivers MUST send all commands for
-a single transaction to the same mongos.
+Mongos Pinning
+~~~~~~~~~~~~~~
+
+Drivers MUST send all commands for a single transaction to the same mongos
+(excluding retries of commitTransaction and abortTransaction).
 
 After the driver selects a mongos for the first command within a transaction,
 the driver MUST pin the ClientSession to the selected mongos. Drivers MUST
-send all commands that are part of the same transaction, including
-commitTransaction and abortTransaction and any retries thereof, to the
-same mongos.
+send all subsequent commands that are part of the same transaction (excluding
+retries of commitTransaction and abortTransaction) to the same mongos.
+
+When to unpin
+^^^^^^^^^^^^^
+
+Drivers MUST unpin a ClientSession when a command within a transaction
+fails with a TransientTransactionError. Transient errors indicate that the
+transaction in question has already been aborted or that the pinned mongos is
+down/unavailable. Unpinning the session ensures that a subsequent
+abortTransaction does not block waiting on a server that is unreachable.
+
+Additionally, Drivers MUST unpin a ClientSession after running any
+commitTransaction or abortTransaction attempt regardless if that attempt
+succeeded or failed. After the initial commit or abort attempt, any mongos can
+satisfy a subsequent retry. Note, when the initial attempt fails with a
+retryable error, the automatic retry attempt MUST perform normal
+server selection to select an available mongos.
 
 Starting a new transaction on a pinned ClientSession MUST unpin the
 session. Additionally, any non-transaction operation using a pinned
 ClientSession MUST unpin the session and the operation MUST perform normal
 server selection.
 
+recoveryToken field
+~~~~~~~~~~~~~~~~~~~
+
+The ``recoveryToken`` field enables the driver to recover the outcome of a
+sharded transaction on a new (or restarted) mongos. [#]_
+
+When a driver runs a command within a transaction, the mongos response
+includes a ``recoveryToken`` field. Drivers MUST track the most recently
+received ``recoveryToken`` field and MUST append this field to any subsequent
+commitTransaction or abortTransaction commands.
```

Comment:
only commitTransaction

Reply:
Updated to only send recoveryToken with commitTransaction.

Reply:
Related question: should drivers attach the recoveryToken *only on retries*? This way when we run the original commit against a mongos that was restarted (or a new mongos) we'll get an immediate error instead of blocking until the transaction is automatically aborted.

Reply:
Currently, mongos will only use the recoveryToken only if it doesn't know who is the coordinator. So it is harmless to attach it on the non-retry cases. Not sure if this will change in the future.

Reply:
I would argue that it is harmful to include it on the original attempt. When we include it on the original attempt and the mongos does not know the coordinator, the mongos will block for 60 seconds and then respond with a transient TransactionAborted error. 

If we do not include it on the original attempt and the mongos does not know the coordinator, the mongos will return a transient error *immediately*. 

I think the second case is much more desirable especially when we consider that the next operations on the session will very likely be retrying the entire transaction (as in the `withTransaction` API).

Reply:
Hmm, I just tested this and found that my assumption that "mongos will return a transient error _immediately_" is not true. Mongos does return an error immediately but the error does not include the `TransientTransactionError` label. The error is:
```
{'ok': 0.0, 'errmsg': 'cannot commit with no participants', 'code': 50940, 'codeName': 'Location50940'}
```

I suppose that the mongos cannot distinguish this case from the case where a client sends a *retry* without a recoveryToken so it cannot add the `TransientTransactionError` label. Is that correct?

Reply:
Given the server behavior here I'm okay with leaving the spec as is (drivers send recoveryToken on all commit attempts). I don't think the behavior of either case is ideal but at least after blocking for 60 seconds the transaction can be retried automatically by `withTransaction`.

I'll open a new server ticket to discuss this problem because I'm not sure drivers can workaround it.

Reply:
Sounds good to me

Reply:
Great, I also think the driver behavior here is correct.  I think the issue is that this section from the server design doc may not have been implemented yet:
> Doing so will never initiate committing the transaction. **Instead, the recovery token in the request will be used to try to abort the transaction if a decision to commit has not already been made**, otherwise to recover the transaction's outcome.

The 60 second waiting period should be removed once the server implements this abort logic.

Reply:
Hm... I think I understand the source of the confusion for about recoveryToken for abortTransasction. The recoveryToken was meant to be used only for commit from the start. What the sentence is saying is that recovery will try to abort the transaction if decision has not been made. The current implementation is it will always try to join the active coordinator. Only if the coordinator no longer exists (because the commit finished/or was aborted) will it try to call abort (just to find out the result, if the response returned error because "transaction already committed", then we know it was committed).

Reply:
Let's continue this conversation on https://jira.mongodb.org/browse/SERVER-39726?